pspmutating.go

```diff
@@ -0,0 +1,133 @@
+package pspmigrator
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-test/deep"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func GetContainerSecurityContexts(podSpec v1.PodSpec) []*v1.SecurityContext {
+	scs := make([]*v1.SecurityContext, 0)
+	for _, c := range podSpec.Containers {
+		scs = append(scs, c.SecurityContext)
+	}
+	return scs
+}
+
+func GetPSPAnnotations(annotations map[string]string) map[string]string {
+	pspAnnotations := make(map[string]string)
+	for ann, val := range annotations {
+		if strings.Contains(ann, "seccomp.security") || strings.Contains(ann, "apparmor.security") {
+			pspAnnotations[ann] = val
+		}
+	}
+	return pspAnnotations
+}
+
+// IsPodBeingMutatedByPSP returns whether a pod is likely mutated by a PSP object. It also returns the difference
+// of the securityContext attribute between the parent controller (e.g. Deployment) and the running pod.
+func IsPodBeingMutatedByPSP(pod *v1.Pod, clientset *kubernetes.Clientset) (mutating bool, diff []string, err error) {
+	diff = make([]string, 0)
+	if len(pod.ObjectMeta.OwnerReferences) > 0 {
+		var owner metav1.OwnerReference
+		for _, reference := range pod.ObjectMeta.OwnerReferences {
+			if reference.Controller != nil && *reference.Controller == true {
+				owner = reference
+				break
+			}
+		}
+		var parentPod v1.PodTemplateSpec
+		if owner.Kind == "ReplicaSet" {
+			rs, err := clientset.AppsV1().ReplicaSets(pod.Namespace).Get(context.TODO(), owner.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, diff, err
+			}
+			parentPod = rs.Spec.Template
+		}
+		if owner.Kind == "DaemonSet" {
+			ds, err := clientset.AppsV1().DaemonSets(pod.Namespace).Get(context.TODO(), owner.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, diff, err
+			}
+			parentPod = ds.Spec.Template
+		}
+		if owner.Kind == "Node" {
+			return false, diff, fmt.Errorf("Pod with ownerReference of kind Node is not supported. OwnerReference of pod %v was %#v", pod.Name, owner)
+		}
+		if diffNew := deep.Equal(GetContainerSecurityContexts(parentPod.Spec), GetContainerSecurityContexts(pod.Spec)); diffNew != nil {
```

Comment:
nit: I'd prefer to use 1st party libraries for this - https://pkg.go.dev/k8s.io/apimachinery/pkg/api/equality#pkg-variables and https://pkg.go.dev/k8s.io/apimachinery@v0.24.3/pkg/util/diff

Reply:
interesting didn't know K8s had a library for stuff like this. I will add this as todo to see if that library will work well enough for this use case too. Looks like this would be the recommended library: github.com/google/go-cmp